File: domain/src/main/java/com/github/anrimian/musicplayer/domain/utils/rx/RxUtils.java
package com.github.anrimian.musicplayer.domain.utils.rx;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Maybe;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.disposables.Disposable;
import io.reactivex.rxjava3.subjects.BehaviorSubject;

public class RxUtils {

    public static <T> Flowable<T> repeatUntilComplete(Maybe<T> maybe) {
        class MaybeCompletedException extends RuntimeException {}

        return maybe.doOnComplete(() -> {
            throw new MaybeCompletedException();
        }).repeat()
                .onErrorResumeNext(throwable -> {
                    if (throwable instanceof MaybeCompletedException) {
                        return Flowable.empty();
                    }
                    return Flowable.error(throwable);
                });
    }

    public static Completable share(Completable completable) {
        return completable.toObservable().share().flatMapCompletable(o -> Completable.complete());
    }

    public static boolean isInactive(@Nullable Disposable disposable) {
        return disposable == null || disposable.isDisposed();
    }

    public static boolean isActive(@Nullable Disposable disposable) {
        return disposable != null && !disposable.isDisposed();
    }

    public static void dispose(@Nullable Disposable disposable,
                               @Nonnull CompositeDisposable compositeDisposable) {
        if (disposable != null && !disposable.isDisposed()) {
            compositeDisposable.remove(disposable);
        }
    }

    public static void dispose(@Nullable Disposable disposable) {
        if (disposable != null && !disposable.isDisposed()) {
            disposable.dispose();
        }
    }

    public static <T> Observable<T> withDefaultValue(BehaviorSubject<T> subject, Creator<T> creator) {
        return Observable.<T>create(emitter -> {
            if (subject.getValue() == null) {
                try {
                    T value = creator.create();
                    if (value != null) {
                        subject.onNext(value);
                    }
                } catch (Exception e) {
                    emitter.onError(e);
                }
            }
        }).mergeWith(subject);
    }

    public static <T> Observable<T> withDefaultValue(BehaviorSubject<T> subject, Single<T> creator) {
        return Observable.<T>create(emitter -> {
            if (subject.getValue() == null) {
                //noinspection ResultOfMethodCallIgnored
                creator.subscribe(subject::onNext, emitter::onError);
            }
        }).mergeWith(subject);
    }

    public static <T> Observable<T> withDefaultValue(BehaviorSubject<T> subject, Maybe<T> creator) {
        return Observable.<T>create(emitter -> {
            if (subject.getValue() == null) {
                //noinspection ResultOfMethodCallIgnored
                creator.subscribe(subject::onNext, emitter::onError);
            }
        }).mergeWith(subject);
    }

    public interface Creator<T> {
        T create();
    }
}

File: domain/src/main/java/com/github/anrimian/musicplayer/domain/utils/rx/RxUtils.kt
package com.github.anrimian.musicplayer.domain.utils.rx

import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.subjects.BehaviorSubject

fun <T: Any> Observable<T>.doOnFirst(action: (T) -> Unit): Observable<T> {
    return take(1)
        .doOnNext { action.invoke(it) }
        .concatWith(skip(1))
}

fun <T : Any> Observable<T>.doOnEvent(completableAction: (T) -> Completable?): Observable<T> {
    return flatMapSingle { o ->
        val s = Single.just(o)
        return@flatMapSingle completableAction(o)?.andThen(s) ?: s
    }
}

fun <T: Any> BehaviorSubject<T>.withDefaultValue(creator: Single<T>): Observable<T> {
    return Observable.create<T> { emitter ->
        if (!this.hasValue()) {
            val d = creator.subscribe(this::onNext, emitter::onError)
            emitter.setDisposable(d)
        }
    }.mergeWith(this)
}

/**
 * Never intended to be unsubscribed
 */
class LazyBehaviorSubject<T : Any>(
    defaultValuesObservable: Observable<T>,
) {

    private val subject = BehaviorSubject.create<T>()
    private val observable = Observable.create<T> { emitter ->
        if (!subject.hasValue()) {
            val d = defaultValuesObservable.subscribe(this::onNext, emitter::onError)
            emitter.setDisposable(d)
        }
    }.mergeWith(subject)
        .replay(1)
        .autoConnect()

    fun getObservable() = observable
    fun onNext(value: T) = subject.onNext(value)
    fun getValue(defaultValue: T) = observable.first(defaultValue)
}